pyteal/stack_frame.py

```diff
@@ -0,0 +1,739 @@
+        if not isinstance(other, PyTealFrame):
+            return False
+
+        return all(
+            [
+                self.frame_info == other.frame_info,
+                self.node == other.node,
+                self.rel_paths == other.rel_paths,
+            ]
+        )
+
+    def spawn(self, other: "PyTealFrame", status: PytealFrameStatus) -> "PyTealFrame":
+        assert isinstance(other, PyTealFrame)
+
+        ptf = PyTealFrame(
+            frame_info=other.frame_info,
+            node=other.node,
+            creator=other.creator,
+            full_stack=other.full_stack,
+            rel_paths=other.rel_paths,
+            parent=self,
+        )
+        ptf._status = status
+
+        return ptf
+
+    def location(self) -> str:
+        return f"{self.file()}:{self.lineno()}" if self.frame_info else ""
+
+    def file(self) -> str:
+        if not self.frame_info:
+            return ""
+
+        path = self.frame_info.filename
+        return os.path.relpath(path) if self.rel_paths else path
+
+    def root(self) -> str:
+        if not self.frame_info:
+            return ""
+
+        path = self.frame_info.filename
+        return path[: -len(self.file())]
+
+    def code_qualname(self) -> str:
+        return (
+            Source.executing(self.frame_info.frame).code_qualname()
+            if self.frame_info
+            else ""
+        )
+
+    def lineno(self) -> int | None:
+        naive_lineno = self.frame_info.lineno if self.frame_info else None
+        if naive_lineno is None:
+            return naive_lineno
+
+        hybrid_line, offset = self._hybrid_w_offset()
+        if hybrid_line.startswith(self.raw_code()):
+            return naive_lineno
+
+        return naive_lineno + offset
+
+    def column(self) -> int:
+        """Provide accurate column info when available. Or 0 when not."""
+        return self.node_col_offset() or 0
+
+    def compiler_generated_reason(self) -> str | None:
+        """
+        None indicates either "unkown" or "not compiler generated".
+        To distinguish between these two usages, call `compiler_generated()` first.
+        """
+        if not self.compiler_generated():
+            return None
+
+        if self.creator._compiler_gen:
+            return PT_GENERATED.FLAGGED_BY_DEV
+
+        for k, v in _PT_GEN.items():
+            if k in self.raw_code():
+                return v
+
+        raise AssertionError(
+            "This should never happen as the call to self.compiler_generated() should have prevented this case."
+        )
+
+    # VARIATIONS ON THE THEME OF SOURCE CODE:
+
+    def raw_code(self) -> str:
+        if self._raw_code is None:
+            self._raw_code = (
+                ("".join(self.frame_info.code_context)).strip()
+                if self.frame_info and self.frame_info.code_context
+                else ""
+            )
+
+        return self._raw_code
+
+    def hybrid_unparsed(self) -> str:
```

Comment:
_hybrid_ stands for "combination". The logic combines code the the comes from the AST, if we successfully get it, or from the file system (`frame_info.code_context`) as a backup. The biggest question mark for the future attempt at removing the reliance on `frame_info` is what's going to happen to this function. Should we simply give up providing source if the AST unparsing fails?

Reply:
I get it, and I agree that the hybrid is a combination of some smart idea for some specific cases, or use file system (`code_context`) as a backup.

> Should we simply give up providing source if the AST unparsing fails?

I mean... if you cannot get `pt_chunk` from `node_source`, then you can only return from `code_context`, right?

I am totally fine with current implementation, but do we want to remove reliance on `frame_info` totally? and for what reason...?

Reply:
Not for this PR. But creating the `FrameInfo`'s is what appears to be the slowest part of the source mapper. And finding the code using AST's `unparse()` doesn't actually require it. So if we aim at the most efficient algorithm that works most of the time, we wouldn't want to use `FrameInfo` and therefore wouldn't have `code_context` as a backup.

Reply:
Sounds very reasonable. I am quite surprised that creating `FrameInfo` object can be hurting the perf, but since we can play with `unparse` from `AST`, then I think we can still get similar info (say line of source).

This give rise 2 concerns, but not in this PR:
- `StackFrame` has `frame_info: FrameInfo` as its instance member, you might have to refactor it a bit
- you might also need to refactor `raw_code` and `_hybrid_impl` a bit for different logic

But not in this PR I think.